pkg/appgateserver/rings.go

```diff
@@ -0,0 +1,135 @@
+	apptypes "github.com/pokt-network/poktroll/x/application/types"
+)
+
+// getRingSingerForAppAddress returns the RingSinger used to sign relays.
+// This method first attempts to get the points of the ring from the cache, if it
+// fails it queries the application module for the points and creates the ring.
+func (app *appGateServer) getRingSingerForAppAddress(ctx context.Context, appAddress string) (*signer.RingSigner, error) {
+	var ring *ring.Ring
+	var err error
+
+	// lock the cache for reading
+	app.ringCacheMutex.RLock()
+	defer app.ringCacheMutex.RUnlock()
+
+	// check if the ring is in the cache
+	points, ok := app.ringCache[appAddress]
+	if !ok {
+		// if the ring is not in the cache, get it from the application module
+		log.Printf("DEBUG: No ring cached for address: %s", appAddress)
+		ring, err = app.getRingForAppAddress(ctx, appAddress)
+	} else {
+		// if the ring is in the cache, create it from the points
+		log.Printf("DEBUG: Ring cached for address: %s", appAddress)
+		ring, err = newRingFromPoints(points)
+	}
+	if err != nil {
+		log.Printf("ERROR: Unable to get ring for address: %s [%w]", appAddress, err)
+		return nil, err
+	}
+
+	// return the ring signer
+	return signer.NewRingSigner(ring, app.signingInformation.SigningKey), nil
+}
+
+// getRingForAppAddress returns the RingSinger used to sign relays. It does so by fetching
+// the latest information from the application module and creating the correct ring.
+// This method also caches the ring's public keys for future use.
+func (app *appGateServer) getRingForAppAddress(ctx context.Context, appAddress string) (*ring.Ring, error) {
+	points, err := app.getDelegatedPubKeysForAddress(ctx, appAddress)
+	if err != nil {
+		return nil, err
+	}
+	return newRingFromPoints(points)
+}
+
+// newRingFromPoints creates a new ring from a slice of points on the secp256k1 curve
+func newRingFromPoints(points []ringtypes.Point) (*ring.Ring, error) {
+	return ring.NewFixedKeyRingFromPublicKeys(ring_secp256k1.NewCurve(), points)
+}
+
+// getDelegatedPubKeysForAddress returns the ring used to sign a message for the given application
+// address, by querying the portal module for it's delegated pubkeys
+func (app *appGateServer) getDelegatedPubKeysForAddress(
+	ctx context.Context,
+	appAddress string,
+) ([]ringtypes.Point, error) {
+	app.ringCacheMutex.RLock()
+	defer app.ringCacheMutex.RUnlock()
+
+	// get the application's on chain state
+	req := &apptypes.QueryGetApplicationRequest{Address: appAddress}
+	res, err := app.applicationQuerier.Application(ctx, req)
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve application for address: %s [%w]", appAddress, err)
+	}
+
+	// create a slice of addresses for the ring
+	ringAddresses := make([]string, 0)
+	ringAddresses = append(ringAddresses, appAddress) // app address is index 0
+	ringAddresses = append(ringAddresses, appAddress) // add app address twice to make the ring size of mininmum 2
```

Comment:
```suggestion
	// TODO_TECHDEBT: We are adding the appAddress twice because a ring signature requires AT LEAST
	// two pubKeys. When the Application acts independently, we use the same key twice, but need to research
	// if ring signatuers can be create with just one key.
	ringAddresses = append(ringAddresses, appAddress) // add app address twice to make the ring size of mininmum 2
```

Reply:
They cannot this is part of the ring library updated comment with the rest tho